substrate-node/pallets/pallet-smart-contract/src/migrations/v12.rs

```diff
@@ -0,0 +1,200 @@
+                )
+            );
+        }
+
+        check_contract_lock_v12::<T>()
+    }
+}
+
+#[cfg(feature = "try-runtime")]
+fn validate_pallet_version<T: Config>(
+    expected_version: types::StorageVersion,
+) -> Result<(), sp_runtime::TryRuntimeError> {
+    let current_version = PalletVersion::<T>::get();
+    debug!("current pallet version: {:?}", current_version);
+    ensure!(
+        current_version >= expected_version,
+        DispatchError::Other("Unexpected pallet version")
+    );
+    Ok(())
+}
+
+pub fn migrate_to_version_12<T: Config>() -> frame_support::weights::Weight {
+    debug!(
+        " >>> Starting contract pallet migration, pallet version: {:?}",
+        PalletVersion::<T>::get()
+    );
+
+    let mut total_reads = 0;
+    let mut total_writes = 0;
+
+    for (contract_id, contract) in Contracts::<T>::iter() {
+        log::debug!("Contract id: {:?}", contract_id);
+
+        if let Some(src_twin) = pallet_tfgrid::Twins::<T>::get(contract.twin_id) {
+            if ContractLock::<T>::contains_key(contract_id) {
+                let (r, w) = migrate_contract_lock::<T>(contract_id);
+                total_reads += r;
+                total_writes += w;
+
+                let (r, w) = remove_all_locks::<T>(&src_twin.account_id);
+                total_reads += r;
+                total_writes += w;
+            } else {
+                log::debug!("ContractLock not found for contract {:?}", contract_id);
+            }
+        } else {
+            log::debug!("Twin not found for contract {:?}", contract_id);
+        }
+    }
+
+    // Set the new storage version
+    PalletVersion::<T>::put(types::StorageVersion::V12);
+    total_writes += 1;
+
+    T::DbWeight::get().reads_writes(total_reads, total_writes)
+}
+
+fn migrate_contract_lock<T: Config>(contract_id: u64) -> (u64, u64) {
+    let mut reads = 0;
+    let mut writes = 0;
+
+    let old_contract_lock = ContractLock::<T>::take(contract_id);
+    reads += 1;
+    writes += 1;
+
+    ContractPaymentState::<T>::insert(
+        contract_id,
+        super::types::v12::ContractPaymentState {
+            standard_reserve: BalanceOf::<T>::zero(),
+            additional_reserve: BalanceOf::<T>::zero(),
+            standard_overdraft: old_contract_lock.amount_locked,
+            additional_overdraft: old_contract_lock.extra_amount_locked,
+            last_updated_seconds: old_contract_lock.lock_updated,
+            cycles: old_contract_lock.cycles,
+        },
+    );
+    writes += 1;
+
+    (reads, writes)
+}
+
+fn remove_all_locks<T: Config>(account_id: &T::AccountId) -> (u64, u64) {
+    let mut reads = 0;
+    let mut writes = 0;
+
+    let locks = pallet_balances::Pallet::<T>::locks(account_id);
+    reads += 1;
+
+    for lock in locks {
+        log::debug!("Removing lock: {:?} for account: {:?}", lock.id, account_id);
+        pallet_balances::Pallet::<T>::remove_lock(lock.id, account_id);
+        reads += 1;
+        writes += 1;
+    }
```

Comment:
Another doubt here...
I am not sure we should remove all locks
Maybe just removing the `GRID_LOCK_ID` lock is enough since this is the one we used in with `ContractLock` storage.

Reply:
First, in the current implementation, locks are only used in the smart contract pallet, so it is safe to remove all locks as we migrate from them.
Second, I meant to remove all locks because I noticed before some leftover locks from an old implementation with different names before using GRID_LOCK_ID.
My account on devnet can serve as an example as well:
5DFkH2fcqYecVHjfgAEfxgsJyoEg5Kd93JFihfpHDaNoWagJ

```
[
  {
    id: 0x0000000000000050
    amount: 631,048,769
    reasons: Misc
  }
  {
    id: 0x000000000000005a
    amount: 640,655,139
    reasons: Misc
  }
  {
    id: 0x0000000000000077
    amount: 613,120,026
    reasons: Misc
  }
  {
    id: 0x000000000000007c
    amount: 615,344,631
    reasons: Misc
  }
  {
    id: 0x000000000000007f
    amount: 619,049,311
    reasons: Misc
  }
  {
    id: 0x0000000000000082
    amount: 634,682,950
    reasons: Misc
  }
  {
    id: 0x0000000000000464
    amount: 8,818,612,335
    reasons: Misc
  }
  {
    id: 0x00000000000002ed
    amount: 5,334,075,143
    reasons: Misc
  }
  {
    id: 0x0000000000000a82
    amount: 841,878
    reasons: Misc
  }
  {
    id: 0x0000000000000613
    amount: 23,429,034
    reasons: Misc
  }
  {
    id: 0x00000000000006d2
    amount: 73,693,350
    reasons: Misc
  }
  {
    id: 0x00000000000002eb
    amount: 47,553,021
    reasons: Misc
  }
  {
    id: 0x00000000000002f0
    amount: 53,892,981
    reasons: Misc
  }
  {
    id: 0x0000000000000dde
    amount: 3,467
    reasons: Misc
  }
  {
    id: gridlock
    amount: 8,819,115,335
    reasons: All
  }
]
```

moreover, I have 881.9 TFT locked with gridlock ID, even though I have no contracts at all.

This is why I am cleaning all the locks.

Reply:
Ok good
Let s keep it like this